Update lesson (#1)
learn-co-curriculum/phase-1-lexical-scoping-readme: README.md

# Lexical Scope

## Learning Goals

- Describe lexical scope as the scope where a variable or function is defined.

> - Examine how lexical scoping informs the scope chain of a function.

Reviewer: Consider updating _Examine_ to something even more specific/measurable such as _Determine_ or _Identify_ or _Recognize_.

## Introduction

In our journey to understand scope, we've learned about the three different
levels, the scope chain, and how JavaScript traverses it. What we haven't
learned is how JavaScript decides which outer scope to place into the scope
chain for a new function.

We have an _idea_ of how, based on what we learned about the chain and nesting
scopes. Anything that a scope is nested within would be considered part of its
outer scope chain.

> However, that doesn't give the whole picture. It doesn't get to the root core of

Reviewer 2: ```suggestion
However, that doesn't give the whole picture. It doesn't get to the root of
```

how JavaScript decides where scope is defined when a variable, function, or code
block is written. This is when the concept of lexical scope comes in.

## Lexical Scope

So, what _is_ this magical lexical scope? Before we break it down, let's look at
the following code snippet:

```js
const currentStatus = "Crocheting a sweater 🧶";

function showStatus() {
  console.log("Current Status:");
  console.log(" ", currentStatus);
}

function changeStatus() {
  const currentStatus = "Cat tangled all my yarn! 🙀";
  showStatus();
}
```

> What do you think will be logged to the console when we run `changeStatus()`?

Reviewer: This is a great example of built in _Check for Understanding_ questions. It doesn't require the instructor to monitor or grade anything, but the students get to monitor their own understanding along the way.

```js
changeStatus();
// => Current Status:
// =>  Crocheting a sweater 🧶
```

Did that catch you by surprise? At first glance, you might think the status
should have been `Cat tangled all my yarn! 🙀`, and you wouldn't be alone. After
all, `currentStatus` was set to that inside the `changeStatus()` function that
we called.

Then, why wasn't that the status that was logged? Lexical scope.

Every variable or function you write in JavaScript has a lexical scope. It is
**the scope in which it was originally defined**. Its home scope. It is what
decides what belongs in the scope chain of any particular variable or function.

Why does this matter? As we saw, variables and functions are meant to be used
throughout your code. They can be called upon in various different places. Or,
in terms of scope, it can be called upon in other scopes.

However, **just because a variable or function is used in the context of a
different scope from where it was defined, doesn't mean its scope chain
changes**. Its chain always remains the same, based on its lexical scope.

> Lets write out a bulletpoint scope chain for our example. Here's the code again:

Reviewer 2: ```suggestion
Let's write out a bulletpoint scope chain for our example. Here's the code again:
```

```js
const currentStatus = "Crocheting a sweater 🧶";

function showStatus() {
  console.log("Current Status:");
  console.log(" ", currentStatus);
}

function changeStatus() {
  const currentStatus = "Cat tangled all my yarn! 🙀";
  showStatus();
}
```

Here's the scope chain for the entire code block:

- Global Scope
  - `currentStatus`
  - `showStatus()`
  - `changeStatus()`
    - `currentStatus` (separate from the global one)

We can see that `showStatus()` and `changeStatus()` are on the same level. They

> can use _eachother_, but they do not have access to anything _within

Reviewer 2: ```suggestion
can use _each other_, but they do not have access to anything _within
```

> eachother's_ inner scopes. In other words, they are not part of eachother's

Reviewer 2: ```suggestion
each other's_ inner scopes. In other words, they are not part of each other's
```

`showStatus()`'s individual scope chain.

```js
const currentStatus = "Crocheting a sweater 🧶";

// Correct nesting!
function changeStatus() {
  const currentStatus = "Cat tangled all my yarn! 🙀";

  function showStatus() {
    console.log("Current Status:");
    console.log(" ", currentStatus);
  }

  showStatus();
}

changeStatus();
// => Current Status:
// =>  Cat tangled all my yarn! 🙀
```

The updated bulletpoint scope chain now looks like:

- Global Scope
  - `currentStatus`
  - `changeStatus()`
    - `currentStatus` (separate from the global one)
    - `showStatus()`

Now, when JavaScript performs variable lookup for `currentStatus` as referenced
in `showStatus()`, it will first look inside its own scope. When it doesn't find
it, it will go up one level on the chain to `changeStatus()`. It will find it,
resolve the variable lookup, and use the value as defined in `changeStatus()`.
It will never reach the global scope version as it doesn't need to go up any
further along the chain.

That's the first way to fix the issue.

The second fix is to **make `changeStatus()` update the globally scoped
`currentStatus` variable** instead of creating a new one in its own scope.

> This avoids nesting all together. In fact, this fix only requires two little

Reviewer 2: ```suggestion
This avoids nesting altogether. In fact, this fix only requires two little
```

> <details><summary>Seeing these two changes, which fix do you personally think is best?</summary>
> <p>As previously cautioned, it's best practice to avoid nesting too many things when you can. Always opt for the simplest solution, when it makes sense. </p>
>
> <p>For our particular example, what if we just wanted to show the status without changing it first? With the nested fix, that wouldn't be possible at all, as the showStatus() function only exists within changeStatus(). We are forced to change the status before we can show it.</p>
>
> <p> Of course, you could nest <i>another</i> function inside changeStatus() that handles updating the status separately. Except now you've created three functions, two nested in one, when you could have just stuck with two separate ones.</p>
>
> <p>Nesting gets messy exponentially quickly. When you don't need to, don't!</p>
> </p></details>

## Conclusion

Lexical scope refers to something's home scope. Much like ourselves, a function
or variable's lexical scope lives in just one place. We can visit friends,
cafés, work, school, wherever - but our address will always remain at home.

> Lexical scope is the same. Functions can be invoked wherever possible, but its

Reviewer 2: ```suggestion
Lexical scope is the same. Functions can be invoked wherever possible, but their
```

> home scope, and therefore its accessible scope chain, will always remain the

Reviewer 2: ```suggestion
home scope, and therefore their accessible scope chain, will always remain the
```

same.

As with all things scope, this too may feel like some nebulous concept. Scope is
not something you consciously do or write. In a way, it just happens. However,
where you write things determines how they're scoped and what other parts of
code they have access to. It is an ever-present concept, always there even when
you're not thinking about it.

As you move through the rest of this course and further on your programming
journey, just knowing what all this scope magic is will help immensely.

When a variable contains an unexpected value, understanding the scope chain will

> save you countless hours of painful debugging. When you're wondering where to

Reviewer: Great example of explaining to adult learners how this lesson is immediately relevant to their work!

declare a function so that it can access the proper variables, your familiarity
with JavaScript's lexical scoping will save the day. When you want to impress
some new friends at a party, hit 'em with a quick lesson on how everything

> defined in JavaScript has a home (well, a lexical scope), and isn't that just

Reviewer: LOL
